### ISSUE TITLE
[no-undef] Ignore type references

### ISSUE THREAD

**Issue by edoardocavazza** [typescript-eslint/typescript-eslint#3939]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "env": {
      "shared-node-browser": true
  },
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "lib": ["ESNext"]
  },
  "rules": {
    "no-undef": ["error"]
  }
}
```

```TS
const isElement = (node: Node): node is Element => node.nodeType === Node.ELEMENT_NODE;
```

```
{
    "compilerOptions": {
        "target": "ESNext",
        "module": "ESNext",
        "lib": ["ESNext", "DOM"],
        "useDefineForClassFields": false,
        "strict": true,
        "strictNullChecks": true,
        "noImplicitAny": true,
        "moduleResolution": "node",
        "esModuleInterop": true,
        "experimentalDecorators": true,
        "skipLibCheck": true,
        "forceConsistentCasingInFileNames": true,
        "types": []
    }
}
```

**Expected Result**

Type references in function signature are typechecked by TS and ignored by ESLint, but the Node reference in the function body should warn in order to honour `shared-node-browser`. Please note that the `Node` constructor may be provided by UI libraries that target both browser and Node for SSR, so it actually can be imported in order to solve the reference problem, while `Element` is used as type only.

**Actual Result**

If the lib.dom is loaded, ESLint does not warn about `Node` usage in function body even if the env is setted to `shared-node-browser`, since the plugin "registers" global variables from lib.dom.

If the lib.dom is NOT loaded, ESLint warns about both type and used references, while TypeScript complains about missing interfaces that are part of the lib.dom but not have not a concrete object to import and use as reference.

**Additional Info**

In the FAQ there is a section about the `no-undef` rule but I was not sure that it covers this case. Would be great to have an option to exclude type references from the list of `no-undef` matches.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.32.0` |
| `@typescript-eslint/parser`        | `4.32.0` |
| `TypeScript`                       | `4.4.3` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `16.5.0` |
